### PR TITLE
Add tmux support to Harpoon

### PR DESCRIPTION
**Pull request by pranavrao145**
This PR introduces integrates Harpoon terminals and tmux.

### New Features
- Adds module `harpoon.tmux`, which implements various functions that can be used to help Harpoon interact with `tmux`. The module implements the following functions for the user:
  - `gotoTerminal`
  - `sendCommand`
  - `clear_all`
  - and pretty much any and all functions implemented by the already implemented `term` module
 - Adds a configuration option that closes all `tmux` windows generated by Harpoon on `nvim` close. This option is disabled by default (meaning any Harpoon-generated tmux windows stay open after `nvim` closes.

### Dev Notes
  - I decided to implement the tmux functionality in a separate module, so that people have the option to choose between `harpoon.term` and `harpoon.tmux`.
  - I've set it up so`harpoon.tmux` basically exactly mimics the `harpoon.term` module, meaning if someone wanted to use `tmux` instead of `term` they would just have to change `harpoon.term` to `harpoon.tmux` in the require statement for their remaps. Same config, no extra work required.
  - This also makes it easier to get rid of the `term` module later on if we need to.

### Issues/Next Steps
  - Term utils: there are a lot of functions in the `term` module used for testing and manipulating the terminal configuration dynamically (`add_cmd`, `rm_cmd`, etc.). I've literally copied and pasted those functions into the `tmux` module so they can be called as `harpoon.tmux.[[ function ]]()`, but I'm honestly not sure if this is necessary. I really think they should be in one module, or they should be refactored and put in an external utils module. I've left them in there for now, but all those changes are in one commit so they can be dropped very easily.
  - Next steps:
    - Jumping back to `nvim` from `tmux`: it would be really nice if there was a quick bind to jump back from `tmux` to `nvim` (like how I could jump from the `nvim` terminal back to my file with a mark). Sure, you could just use the `tmux` binds to find the session with `nvim` in it, but where's the fun in that? I have absolutely no idea how to implement this (or if it's even possible), but if someone wants to do that and knows how, be my guest.
    - Docs: assuming people like what they see here, I can write up the docs to this as well later on. Just wanted to finalize the API first!
 
Alright that's pretty much it, sorry that was so long. Please note this is my first time working with `lua`, the `nvim` API, or even on open source, so if there's anything I missed in terms of whitespace, conventions, etc., please let me know and I can fix it. 

Thanks!

### CODE REVIEW

**Comment by ThePrimeagen** [comment]
I am going to start using this shortly!  Could you fix the small conflict? @pranavrao145 .  Once that is done, i'll start using this branch and see how it goes ;)

**Comment by pranavrao145** [comment]
There we go, should be good!

About that one issue I discovered, after testing for about two weeks I've started to realize it's actually an issue with the way `plenary` interacts with `tmux`, rather than the code itself. For some reason, `plenary` jobs are not always receiving the same output that the regular shell is. I've tried various solutions, and while some of them are better, they're still not perfect.

If anyone has any ideas about what to do about this, that would be awesome, because it really has me stumped.

**Comment by pranavrao145** [comment]
@ThePrimeagen, after a lot of experimentation, I think I've fixed it! With the new changes I've put in the first commit, the codebase is cleaner and the issue mentioned above no longer occurs. Check it out and let me know what you think!

**Comment by ThePrimeagen** [comment]
@pranavrao145 i'll merge it once you fix the styles

File: lua/harpoon/tmux.lua [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+    if window_handle then
+        window_exists = terminal_exists(window_handle.window_id)
+    end
+
+    if not window_handle or not window_exists then
+        local window_id = create_terminal()
+
+        if window_id == nil then
+            return
```
[the remark this answers is not in the document]
Comment (pranavrao145) [reply lua/harpoon/tmux.lua:83]: Sure, I left it like this so it's consistent with the `term` module, but I can add in error reporting for both of them.

File: lua/harpoon/tmux.lua [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        end
+    end
+    return M.get_length() + 1
+end
+
+M.gotoTerminal = function(idx)
+    log.trace("tmux: gotoTerminal(): Window:", idx)
+    local window_handle = find_terminal(idx)
+
+    local _, _, _ = utils.get_os_command_output({
```
[the remark this answers is not in the document]
Comment (pranavrao145) [reply lua/harpoon/tmux.lua:110]: No idea why I did that, I'll fix that.

File: lua/harpoon/tmux.lua [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        }, vim.loop.cwd())
+    end
+end
+
+M.clear_all = function()
+    log.trace("tmux: clear_all(): Clearing all tmux windows.")
+
+    for _, window in pairs(tmux_windows) do
+        -- Delete the current tmux window
+        local _, _, _ = utils.get_os_command_output({
```
[the remark this answers is not in the document]
Comment (pranavrao145) [reply lua/harpoon/tmux.lua:149]: Same thing, I'll delete these.

**Comment by pranavrao145** [comment]
Sure, I'll fix those things and the style, as well as add some docs.

**Comment by ThePrimeagen** [comment]
@pranavrao145 I am going to give this a real try.  See how it works

File: lua/harpoon/tmux.lua [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+    local window_id
+
+    -- Create a new tmux window and store the window id
+    local out, ret, _ = utils.get_os_command_output({
+        "tmux",
+        "new-window",
+        "-P",
+        "-F",
+        "#{pane_id}",
```
Comment (tg0h) [inline lua/harpoon/tmux.lua:28]: @pranavrao145 shouldn't this be window_id instead ?

I notice that the go to terminal window command mostly works, however if i start creating panes, the go to terminal command starts becoming flaky

**Pull request merged**